File: petclinic-java-gradle-react-tailwind-ts/src/main/java/dev/pollito/petclinic_java_gradle_react_tailwind_ts/pet/PetMapper.java
package dev.pollito.petclinic_java_gradle_react_tailwind_ts.pet;

import dev.pollito.petclinic_java_gradle_react_tailwind_ts.generated.model.PetDTO;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.owner.Owner;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.owner.OwnerRepository;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.type.Type;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.type.TypeRepository;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.NotFoundException;
import org.mapstruct.AfterMapping;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants;
import org.mapstruct.MappingTarget;
import org.mapstruct.ReportingPolicy;
import org.springframework.beans.factory.annotation.Autowired;

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING, unmappedTargetPolicy = ReportingPolicy.IGNORE)
public abstract class PetMapper {

    @Autowired
    protected TypeRepository typeRepository;

    @Autowired
    protected OwnerRepository ownerRepository;

    @Mapping(target = "type", expression = "java(pet.getType() == null ? null : pet.getType().getId())")
    @Mapping(target = "owner", expression = "java(pet.getOwner() == null ? null : pet.getOwner().getId())")
    public abstract PetDTO map(Pet pet);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "type", ignore = true)
    @Mapping(target = "owner", ignore = true)
    public abstract Pet map(PetDTO petDTO);

    @AfterMapping
    protected void afterMap(PetDTO petDTO, @MappingTarget Pet pet) {
        final Type type = petDTO.getType() == null ? null
                : typeRepository.findById(petDTO.getType())
                        .orElseThrow(() -> new NotFoundException("type not found"));
        pet.setType(type);
        final Owner owner = petDTO.getOwner() == null ? null
                : ownerRepository.findById(petDTO.getOwner())
                        .orElseThrow(() -> new NotFoundException("owner not found"));
        pet.setOwner(owner);
    }
}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/pet/PetMapper.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.pet

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.owner.OwnerRepository
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.type.TypeRepository
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.NotFoundException
import org.mapstruct.AfterMapping
import org.mapstruct.Context
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingConstants
import org.mapstruct.MappingTarget
import org.mapstruct.ReportingPolicy

@Mapper(
    componentModel = MappingConstants.ComponentModel.SPRING,
    unmappedTargetPolicy = ReportingPolicy.IGNORE,
)
interface PetMapper {

    @Mapping(
        target = "type",
        ignore = true,
    )
    @Mapping(
        target = "owner",
        ignore = true,
    )
    fun updatePetDTO(pet: Pet, @MappingTarget petDTO: PetDTO): PetDTO

    @AfterMapping
    fun afterUpdatePetDTO(pet: Pet, @MappingTarget petDTO: PetDTO) {
        petDTO.type = pet.type?.id
        petDTO.owner = pet.owner?.id
    }

    @Mapping(
        target = "id",
        ignore = true,
    )
    @Mapping(
        target = "type",
        ignore = true,
    )
    @Mapping(
        target = "owner",
        ignore = true,
    )
    fun updatePet(
        petDTO: PetDTO,
        @MappingTarget pet: Pet,
        @Context typeRepository: TypeRepository,
        @Context ownerRepository: OwnerRepository,
    ): Pet

    @AfterMapping
    fun afterUpdatePet(
        petDTO: PetDTO,
        @MappingTarget pet: Pet,
        @Context typeRepository: TypeRepository,
        @Context ownerRepository: OwnerRepository,
    ) {
        val type = if (petDTO.type == null) {
            null
        } else {
            typeRepository.findById(petDTO.type!!)
                .orElseThrow { NotFoundException("type not found") }
        }
        pet.type = type
        val owner = if (petDTO.owner == null) {
            null
        } else {
            ownerRepository.findById(petDTO.owner!!)
                .orElseThrow { NotFoundException("owner not found") }
        }
        pet.owner = owner
    }
}
